pymodm/dereference.py

```diff
@@ -117,9 +119,13 @@ def _attach_objects_in_path(container, document_map, fields, key, field):
             not isinstance(value, field.related_model)):
         # value is reference id
         meta = field.related_model._mongometa
-        container[key] = _get_reference_document(document_map,
-                                                 meta.collection_name,
-                                                 meta.pk.to_mongo(value))
+        dereferenced_document = _get_reference_document(
+            document_map, meta.collection_name, meta.pk.to_mongo(value))
+        try:
+            container.set_mongo_value(key, dereferenced_document)
+        except AttributeError:
+            container[key] = field.related_model.from_document(
+                dereferenced_document)
```

Comment:
Because if we are here, `container` is either a list or a dict (corresponding to a `ListField` or `DictField` field). Previously, when we fetched values the `__get__` descriptors always invoked the `to_python` logic which resulted in this document being converted to a model at the time of attribute access. Now that we want to call `to_python` only the requisite number of times (and not every single time we access an attribute) we must create this document at the time we dereference a field. There are obviously performance implications for this as in the older pattern you did not waste CPU cycles creating model instances that you would never use (model instances were populated on-demand from the document dict), whereas now you translate all docs into the corresponding models. I couldn't figure out a performant way to get the correct behavior without touching various parts of the code so it will be an activity that can be taken on in the future.